firebase-firestore/src/main/java/com/google/firebase/firestore/remote/AbstractStream.java

```diff
@@ -290,18 +294,26 @@ private void close(State finalState, Status status) {
     if (code == Code.OK) {
       // If this is an intentional close ensure we don't delay our next connection attempt.
       backoff.reset();
-
     } else if (code == Code.RESOURCE_EXHAUSTED) {
       Logger.debug(
           getClass().getSimpleName(),
           "(%x) Using maximum backoff delay to prevent overloading the backend.",
           System.identityHashCode(this));
       backoff.resetToMax();
-
     } else if (code == Code.UNAUTHENTICATED) {
       // "unauthenticated" error means the token was rejected. Try force refreshing it in case it
       // just expired.
       firestoreChannel.invalidateToken();
+    } else if (code == Code.UNAVAILABLE) {
+      // These exceptions are thrown when the gRPC stream is closed with an connection error. For
+      // these cases, we need to use a new connection for the next connection attempt, which is
+      // done by marking the underlying channel as idle.
+      if (forceNewConnection
+          || status.getCause() instanceof java.net.ConnectException
```

Comment:
I'm not 100% sure this is okay to do...

It looks like ConnectException usually refers to "Connection Refused" which means the backend is potentially down or intentionally trying to shed traffic (e.g. to mitigate a DoS attack).  So we should probably stick with our usual backoff delay.

Were you seeing this exception in your testing?  Or have some other specific reason to include it?

Reply:
In my testing, I noticed that we get this message sometimes when the app starts in an 100% packet loss state, or when grpc shuts down the connection after running its own backoff (which technically shouldn't happen anymore after we add the connectivity timer). 

The main reason to check for ConnectException is to help the user reconnect faster if they start their app in a 100% packet loss state. Without it, if the user hits the maximum backoff delay before regaining connectivity, it could take up to DEFAULT_BACKOFF_MAX_DELAY (1 min) before making the next connection attempt.

However, protecting against DoS attacks seems like a much more important feature than this, so I'll remove the `java.net.ConnectException` check and only keep the `UnknownHostException` one.

Reply:
I don't think the conclusion here is right.

The origin of a status with an exception like this is always local, so if we get a `ConnectException` we're failing to connect _to the GFE_. Any failure past that point would result in a remote status like a bad gateway or similar HTTP response, which gRPC surfaces differently.

While we want to be generous with avoiding fast retries against the Firestore backend, the likelihood that the GFE would be spilling our load this way is so vanishingly small that we can just treat this as a certainty that the local network is to blame. Or in other words, I can't see any case where retrying faster after seeing a ConnectException would cause Firestore any distress.

Reply:
I admit I don't have any specific knowledge of GFE load shedding behavior, etc. and am happy to defer to you.

Reply:
Reverted back to catch `ConnectException` as well.